rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,404 @@
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 3.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 3:** Architecture of a user uploading from their cloud to our cloud.
+
+##### 1. Get Share Link
+
+The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or submission of files.
+
+##### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery.
+
+###### How to download from a shared link
+
+- [Dropbox](https://help.dropbox.com/files-folders/share/force-download)
+- [Google Drive](https://www.labnol.org/internet/direct-links-for-google-drive/28356/)
+
+##### 4. _Optional_ Queue
+
+This part is optional and depends on the load expected and server type selected. The queue stores the links that need to
+be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
+It also gives us the ability to retry a download if one of the servers fails to download or upload the file.
+
+###### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- The CSP experience and outage
+- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+##### 5. Start an Instance to Download File
+
+An EC2 instance, docker container, or AWS lambda can be used to complete this task.
+
+##### 6. Stream File
+
+The shared link is used to stream the file from one cloud to another. The most time will be spent streaming the file 
+from the CSP to S3. Most CSPs offer a way to validate the integrity of a download
+by providing a hash of the data, for example [Dropbox Content Hash](https://www.dropbox.com/developers/reference/content-hash).
+As the data is streaming from the CSP, the hash of the file will be calculated as it arrives, and again as it's 
+uploaded to S3. The [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
```

Comment:
Note that checksum validation does not happen in S3 for multi-part uploads. Or, at least, it does not happen the way you expect. Rather than getting an md5sum back you get a sum of the md5sum's of each part.

This is an issue that idseq has dealt with if you want more context.

Reply:
Did they resolve it?

Back in DCP land, we used https://github.com/HumanCellAtlas/checksumming_io to solve this problem.